plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/InteractiveLayeredGraphVisitor.java

```diff
@@ -0,0 +1,453 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Kiel University and others.
+ * 
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ * 
+ * SPDX-License-Identifier: EPL-2.0  
+ *******************************************************************************/
+package org.eclipse.elk.alg.layered;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.eclipse.elk.alg.layered.options.CycleBreakingStrategy;
+import org.eclipse.elk.alg.layered.options.LayerConstraint;
+import org.eclipse.elk.alg.layered.options.LayeredOptions;
+import org.eclipse.elk.alg.layered.options.LayeringStrategy;
+import org.eclipse.elk.core.math.KVector;
+import org.eclipse.elk.core.options.CoreOptions;
+import org.eclipse.elk.core.options.Direction;
+import org.eclipse.elk.core.util.IGraphElementVisitor;
+import org.eclipse.elk.graph.ElkConnectableShape;
+import org.eclipse.elk.graph.ElkEdge;
+import org.eclipse.elk.graph.ElkGraphElement;
+import org.eclipse.elk.graph.ElkNode;
+import org.eclipse.elk.graph.ElkPort;
+
+/**
+ * Graph visitor which visits only the root node and recursively steps through the graph
+ * to set interactive options and pseudo coordinates for the layered algorithm.
+ * Interactive layout works via two consecutive layout runs.
+ * The first one is a "normal" run. The second one will is configured using this graph visitor.
+ * The graph visitor recursively sets the interactive strategies for all phases and assigns pseudo positions to all
+ * node based on their last position and the desired layer and position in the layer expressed by the
+ * {@code LayerChoiceConstraint} and {@code PositionChoiceConstraint}.
+ * The pseudo position have to represent layers and the ordering in the layer.
```

Comment:
At some point, it would probably be nice to give an example of how to actually use this.

Reply:
I can put some examples in the elk-models repo. Where do you think is an appropriate location for this?

Reply:
I don't mean graphs, I mean code examples. The `DiagramLayoutEngine` needs to be called in a certain way to make this stuff work, I think it would be beneficial to include a simple code example here. Similar to what the [API docs of the `Scanner` class](https://docs.oracle.com/en/java/javase/13/docs/api/java.base/java/util/Scanner.html) do, for example.

If that also entails adding models to the models repository, they could be placed in `/tests/layered/interactive`?

Reply:
I guess this is rather a KLighD concern isn't it? Usually your synthesis or a general synthesis should add a `CompoundGraphVisitor` with the desired interactive graph visitors as an additional layout run. However, I am against moving the whole thing into KLighD instead, since it concerns the layout algorithms.

Reply:
KLighD is but one consumer of the `DiagramLayoutEngine`. If this thing is in ELK, its usage should be properly documented in ELK. 🙂